pymeasure/instruments/lecroy/lecroyT3DSO1204.py

```diff
@@ -0,0 +1,1151 @@
+        - "type": normal, peak detect, average, high resolution (str)
+        - "requested_points": number of data points requested by the user (int)
+        - "sampled_points": number of data points sampled by the oscilloscope (int)
+        - "transmitted_points": number of data points actually transmitted (optional) (int)
+        - "memory_size": size of the oscilloscope internal memory in bytes (int)
+        - "sparsing": sparse point. It defines the interval between data points. (int)
+        - "first_point": address of the first data point to be sent (int)
+        - "source": source of the data : "C1", "C2", "C3", "C4", "MATH".
+        - "type":  type of data acquisition. Can be "normal", "peak", "average", "highres"
+        - "average": average times of average acquisition
+        - "sampling_rate": sampling rate (it is a read-only property)
+        - "grid_number": number of horizontal grids (it is a read-only property)
+        - "status": acquisition status of the scope. Can be "stopped", "triggered", "ready",
+        "auto", "armed"
+        - "xdiv": horizontal scale (units per division) in seconds
+        - "xoffset": time interval in seconds between the trigger event and the reference position
+        - "ydiv": vertical scale (units per division) in Volts
+        - "yoffset": value that is represented at center of screen in Volts
+        """
+        vals = self.values("WFSU?")
+        preamble = {
+            "sparsing": vals[vals.index("SP") + 1],
+            "requested_points": vals[vals.index("NP") + 1],
+            "first_point": vals[vals.index("FP") + 1],
+            "transmitted_points": None,
+            "source": self.waveform_source,
+            "type": self.acquisition_type,
+            "sampling_rate": self.acquisition_sampling_rate,
+            "grid_number": self._grid_number,
+            "status": self.acquisition_status,
+            "memory_size": self.memory_size,
+            "xdiv": self.timebase_scale,
+            "xoffset": self.timebase_offset
+        }
+        preamble["average"] = self.acquisition_average if preamble["type"][0] == "average" else None
+        strict_discrete_set(self.waveform_source, ["C1", "C2", "C3", "C4", "MATH"])
+        preamble["sampled_points"] = self.acquisition_sample_size(self.waveform_source)
+        if self.waveform_source == "MATH":
+            preamble["ydiv"] = self.math_vdiv
+            preamble["yoffset"] = self.math_vpos
+        else:
+            preamble["ydiv"] = self.ch(self.waveform_source).scale
+            preamble["yoffset"] = self.ch(self.waveform_source).offset
+        return preamble
+
+    def _digitize(self, source, requested_bytes=None):
+        """ Acquire waveforms according to the settings of the acquire commands.
+        Note.
+        If the requested number of bytes is not specified, the default chunk size is used,
+        but in such a case it cannot be quaranteed that the message is received in its entirety.
```

Comment:
As far as I understand [PyVISA's `read_raw`](https://github.com/pyvisa/pyvisa/blob/07b32b7697d15d2a8291471995d5958f21a9d3be/pyvisa/resources/messagebased.py#L388) method reads one chunk after the other, until nothing remains to be read. Therefore, the chunck_size should not matter.

Reply:
That is what I would expect too, but I assure you that it is not what happens. At least with our setup. I admit that the chunk resizing dance is ugly but it is the only way I could get it working.

Reply:
Another source of complexity is the fact that I strive to cover any edge case. Especially when number of requested points, sparsing and averaging are specified at the same time.

Reply:
Finally I found very frustrating that both in the programming guide and in the SCPI API, there is no reliable way to calculate or query the actual number of data points that are going to be transmitted. The only way is to request the waveform and parse the header. That is why I make the preliminary call of the digitize method at line 861.